x/gamm/keeper/joinpool_test.go

```diff
@@ -0,0 +1,149 @@
+package keeper_test
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/osmosis-labs/osmosis/v7/x/gamm/pool-models/balancer"
+	"github.com/osmosis-labs/osmosis/v7/x/gamm/types"
+)
+
+func (suite *KeeperTestSuite) TestRandomizedJoinPoolExitPoolInvariants() {
+	type testCase struct {
+		initialTokensDenomIn  int64
+		initialTokensDenomOut int64
+
+		percentRatioMin int64
+
+		joinDenomInAmt  int64
+		joinDenomOutAmt int64
+	}
+
+	const (
+		denomOut = "denomOut"
+		denomIn  = "denomIn"
+	)
+	joinErrCount := 0
+	exitErrCount := 0
+
+	now := time.Now().Unix()
+	rng := rand.NewSource(int64(now))
+	fmt.Printf("Using random source of %d\n", now)
+
+	// generate test case with randomized initial assets and join/exit ratio
+	newCase := func() (tc *testCase) {
+		tc = new(testCase)
+		tc.initialTokensDenomIn = (rng.Int63() % 1_000_000) + 100
+		tc.initialTokensDenomOut = (rng.Int63() % 1_000_000) + 100
+
+		// 1%~200% of initial assets
+		percentRatioIn := rng.Int63()%200 + 1
+		usePerfectRatio := rng.Int63()%2 == 0
+		tc.joinDenomInAmt = tc.initialTokensDenomIn * percentRatioIn / 100
+		percentRatioOut := percentRatioIn
+		if !usePerfectRatio {
+			percentRatioOut = rng.Int63()%200 + 1
+		}
+		tc.joinDenomOutAmt = tc.initialTokensDenomOut * percentRatioOut / 100
+
+		tc.percentRatioMin = percentRatioIn
+		if percentRatioOut < percentRatioIn {
+			tc.percentRatioMin = percentRatioOut
+		}
+
+		return tc
+	}
+
+	// create pool with randomized initial token amounts
+	// and randomized ratio of join/exit
+	createPool := func(tc *testCase) (poolId uint64) {
+		poolAssetOut := balancer.PoolAsset{
+			Token:  sdk.NewInt64Coin(denomOut, tc.initialTokensDenomOut),
+			Weight: sdk.NewInt(5),
+		}
+
+		poolAssetIn := balancer.PoolAsset{
+			Token:  sdk.NewInt64Coin(denomIn, tc.initialTokensDenomIn),
+			Weight: sdk.NewInt(5),
+		}
+
+		return suite.PrepareBalancerPoolWithPoolAsset([]balancer.PoolAsset{poolAssetOut, poolAssetIn})
+	}
+
+	// joins with predetermined ratio
+	joinPool := func(poolId uint64, tc *testCase) {
+		sender := suite.TestAccs[1]
+		tokensIn := sdk.Coins{
+			sdk.NewInt64Coin(denomIn, tc.joinDenomInAmt),
+			sdk.NewInt64Coin(denomOut, tc.joinDenomOutAmt),
+		}
+		suite.FundAcc(sender, tokensIn)
+
+		initShares, _ := sdk.NewIntFromString("100000000000000000000")
+		sharesWanted := initShares.Mul(sdk.NewInt(tc.percentRatioMin)).QuoRaw(100)
+		ctx, writeFn := suite.Ctx.CacheContext()
+		err := suite.App.AppKeepers.GAMMKeeper.JoinPoolNoSwap(ctx, sender, poolId, sharesWanted, tokensIn)
```

Comment:
I set a breakpoint [on this line ](https://github.com/osmosis-labs/osmosis/blob/ddcb653425e163d7ac1716dba0181a629f15dc28/x/gamm/pool-models/balancer/amm.go#L330)in `CalcJoinPoolShares` but it was never hit. Is that expected?

I'm guessing that this is because this is `JoinPoolNoSwap` where we get "maximal no swap amount" [here](https://github.com/osmosis-labs/osmosis/blob/ddcb653425e163d7ac1716dba0181a629f15dc28/x/gamm/keeper/pool_service.go#L195) so it is practically impossible to reach the uneven case.

Reply:
we were hitting that line previously tho right?

Reply:
We were when we had a randomized test on the amm level (`amm_test.go`) but it got moved for some reason in https://github.com/osmosis-labs/osmosis/pull/1781

@ValarDragon what is the reason for that?